go/login/server/server.go

```diff
@@ -0,0 +1,299 @@
+	return func(srv *LoginServer) error {
+		srv.userHeader = s
+		return nil
+	}
+}
+
+//----- code onwards will be deprecated as soon as integration with go-glome-login is done -----
+func (s *LoginServer) newRouter() *mux.Router { // Review Error Message
+	r := mux.NewRouter().UseEncodedPath()
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		params := mux.Vars(r)
+
+		user := r.Header.Get(s.userHeader)
+
+		if err := handleVersion(params["V"]); err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		dialog, tag, err := s.handleHandshakeV1(params["handshake"])
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		hostID, err := readHost(params["hostID"])
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		action, err := readAction(params)
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		msg, err := readMessage(hostID, action)
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		if err = verifyReceivedTag(dialog, tag, msg); err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		hostIDType := ""
+		s.printToken(w, dialog, hostID, hostIDType, action, msg, user)
+	}
+
+	r.Path("/v{V:[0-9]+}/{handshake}/{hostID}/").HandlerFunc(handler)
+	r.Path("/v{V:[0-9]+}/{handshake}/{hostID}/{action:[a-zA-Z0-9+/=%]+}/").
+		HandlerFunc(handler)
+
+	return r
+}
+
+func handleVersion(V string) error {
+	switch i, err := strconv.Atoi(V); {
+	case err != nil:
+		return err
+	case i != 1:
+		return ErrVersionNotSupported
+	}
+	return nil
+}
+
+func (s *LoginServer) handleHandshakeV1(h string) (*glome.Dialog, []byte, error) {
+	data, err := base64.URLEncoding.DecodeString(h)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(data) < 32 { //handshake must have 66 bytes (2 prefixtype+prefix7, 32 eph, 32 prefixN)
+		return nil, nil, ErrFailedHandshake
+	}
+
+	prefixtype := data[0] >> 7 // Read first bit
+	if prefixtype != 0 {
+		return nil, nil, ErrInvalidPrefixType
+	}
+
+	prefix7 := (data[0] & 127) // Read first byte but fist bit
+	server, found := s.Keys.Read(prefix7)
+	if !found {
+		return nil, nil, ErrInvalidPrefix7
+	}
+
+	client, err := glome.PublicKeyFromSlice(data[1:33])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dialog, err := server.TruncatedExchange(client, 1)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return dialog, data[33:], nil // PrefixN currently stores the message tag.
+}
+
+func readHost(hostID string) (string, error) {
+	return url.QueryUnescape(hostID)
+}
+
+func readAction(params map[string]string) (string, error) {
+	action, ok := params["action"]
+	if !ok {
+		return "", nil
+	}
+
+	return url.QueryUnescape(action)
+}
+
+func readMessage(hostID string, action string) ([]byte, error) {
+	msgAux, err := url.QueryUnescape(hostID + "/" + action)
```

Comment:
action is not escaped, isn't it? According to the [first test vector](https://github.com/google/glome/blob/master/docs/glome-login.md#vector-1).